### ISSUE TITLE
Enforce unique test titles

### ISSUE THREAD

**Issue by sindresorhus** [avajs/ava#517]
From Gitter:

> @sindresorhus: @jamestalmage @vdemedes I had this thought yesterday. Should be disallow duplicate titles in tests? Or rather, is there any good reason to allow it? Duplicate titles makes it really hard to look at the output and see what actually failed. We're already being opinionated on many other things. Open to thoughts from anyone else too btw.

> @jamestalmage: Perhaps just a warning? If I copy/paste a test in —watch mode make a few changes withouth changing the test title and save, I shouldn’t see red. Maybe some yellow though
I think the warning would be sufficient motivation to change the title. Especially for fans of the minimal reporter

Input welcome! :)

**Comment by vdemedes** [comment]
While this is an interesting idea, I'm wondering how it will affect user experience. Are we just going to show an error, instead of running perfectly fine tests?

**Comment by sindresorhus** [comment]
As James pointed out, it could just be a warning.

**Comment by vdemedes** [comment]
Ok, in that case I think the best place for this feature is to be in that ava linter we talked about.

**Comment by sindresorhus** [comment]
That will only benefit people using ESLint and our plugin, though. If we can do things directly in AVA without any negative side-effects, I think we should.

**Comment by jamestalmage** [comment]
> If we can do things directly in AVA without any negative side-effects, I think we should

I agree.

**Comment by jamestalmage** [comment]
@jarettmeyer on Gitter:

> I have a question about #517 . I am thinking of adding this check to TestCollection's add function.
Next, a test needs to be able to return a warning, instead of an error. As I can tell, the runner does not support returning warnings, only errors. Is there another place I should look for warnings?

I think we have two options:

 1. Create a new custom `ipc` event, especially for warnings. This would involve adding appropriate Logger hooks, etc.

 2. Just print warnings to stdout - it will get displayed by AVA.

I think `1`.

**Comment by jarrettmeyer** [comment]
Agreed: #1. If you print to directly to stdout, you can't really suppress it with a minimalist reporter or promote it to an error with eslint.

**Comment by jarrettmeyer** [comment]
Should this check for duplicate names within a single test file, or across all test files?

**Comment by vdemedes** [comment]
In my opinion, we should postpone or completely ignore that feature for now. It kind of feels lately, that we add more bloat to AVA, instead of fixing/improving monsters (perf, babel, no browser support - off the top of my head) we have at the moment.

**Comment by jamestalmage** [comment]
> It kind of feels lately, that we add more bloat to AVA, instead of fixing/improving monsters

I can get behind that sentiment.

**Comment by sindresorhus** [comment]
Agreed. I'm just opening issues when I have ideas. It's not like we have to implement everything now. This is btw what the `blocked` (suggestions for a better title for it?) label is for. I do think this should land before 1.0.0, though, as it's a breaking change.

**Comment by jarrettmeyer** [comment]
I can put this on the back-burner. However, I am almost complete with a first pass. My suggestions for this feature.

1. Add a "warning" event type. Similar to "error" but with a different name. This gets wired all the way through ava, from the initial emit, through `fork`, `api`, `logger`, and finally to the reporters.
2. Add a `--werror` flag to the cli. Stolen from gcc, the `--werror` flag treats warnings as errors.
3. A duplicate test title emits a warning event. Currently, this warning is being emitted from `TestCollection`'s `add()` function, but that only checks that the test title is unique within a single file, not across all files.

**Comment by sindresorhus** [comment]
> Add a --werror flag to the cli. Stolen from gcc, the --werror flag treats warnings as errors.

:-1: This will already be handled by: https://github.com/sindresorhus/eslint-plugin-ava/issues/7

**Comment by jamestalmage** [comment]
I take `blocked` to mean "some other issue is holding this up". I would just mark this `low-priority`.

> I do think this should land before 1.0.0, though, as it's a breaking change.

It's not a breaking change if it's just a warning.

**Comment by novemberborn** [comment]
I'm not convinced AVA should be keeping track of all test titles and warn you when there are duplicates. Assuming test failures come with stack traces they'll also come with line numbers, letting you find the offending assertion and thus test.

The linter rule seems like a sufficient solution.

**Comment by sindresorhus** [comment]
@novemberborn Yeah, let's just go with the ESLint plugin for this.
